File: EXPORT.py
'''
Author: 
    aarontburn (https://github.com/aarontburn)
Title:
    Module Export Script
Purpose:
    Properly export developed module to use with parent application
        (https://github.com/aarontburn/modules)

Repository: 
    https://github.com/aarontburn/modules-export-script
    
Usage: 
    'npm run export', or 'python node_modules/modules-export-script/EXPORT.py'
    
Expected Result: 
    In the root directory, a directory 'output/' will be created containing required files for the module.
'''

import os
import pathlib
import shutil
import json
import time

'''The path of the root directory of this module.'''
PWD: str = str(pathlib.Path(__file__).parent.parent.parent.resolve())

def locateModule() -> str:
    '''
    Scans "src" folder for a directory containing "ref.DAT", which should
    be located in the module directory.
    
    Returns the name of the directory.
    '''
    src_path: str = PWD + "/src/"
    for file in os.listdir(src_path):
        file_path = src_path + file
        if os.path.isdir(file_path) and os.listdir(file_path).count("ref.DAT") > 0:
            return file

'''The name of the directory containing the module to export.'''
FOLDER_NAME: str = locateModule()

'''The path of the output directory in the output folder.'''
OUTPUT_FOLDER_PATH: str = PWD + "/output/" + FOLDER_NAME + "/"

'''The path to the node_modules directory in the output folder.'''
NODE_MODULES_PATH: str = PWD + "/node_modules"


def main() -> None:
    ''' Entry point. '''
    shutil.rmtree(OUTPUT_FOLDER_PATH, ignore_errors=True)
    
    createDirectories()
    copyFiles()
    checkAndCopyDependencies()
    print("\n\tFINISHED BUNDLING MODULE\n")
    


def createDirectories() -> None:
    '''Creates folders required for the export.'''
    
    def mkdir(directoryName: str) -> None:
        '''Helper function to create directories.'''
        try:
            os.makedirs(directoryName)
            print("Creating folder: " + directoryName)
        except FileExistsError:
            print("Removing existing directory: " + directoryName)
        
    print("\n\tCREATING FOLDERS\n")
    mkdir(OUTPUT_FOLDER_PATH)
    mkdir(OUTPUT_FOLDER_PATH + "module_builder")
    mkdir(OUTPUT_FOLDER_PATH + "node_modules")


def copyFiles() -> None:
    '''Copies files into the output folder.'''
    
    print("\n\tCOPYING FILES\n")
    path: str = PWD + "/src/" + FOLDER_NAME + "/"

    for file in os.listdir(path):
        full_file_path = path + file
        output_file_path = OUTPUT_FOLDER_PATH + file
        
        if (os.path.isfile(full_file_path)):
            print("Copying '" + full_file_path + "' to output folder ('" + output_file_path + "')")
            shutil.copyfile(full_file_path, output_file_path)


def checkAndCopyDependencies() -> None:
    '''Checks the dependencies and bundles them with output folder.'''
    
    file_text: str = ''
    with open("package.json", "r") as file:
        file_text = file.read()

    if file_text == '':
        print("Could not open package.json")
        return

    package_json: any = json.loads(file_text)
    dependencies: any = package_json['dependencies']

    if (len(dependencies.keys()) > 0):
        print("\n\tBUNDLING DEPENDENCIES\n")


    node_modules: list[str] = os.listdir(NODE_MODULES_PATH)
    for dependency_name in dependencies.keys():
        # version = dependencies[dependency]
        try:
            node_modules.index(dependency_name)
        except ValueError:
            print(dependency_name + " was not found in node_modules. Skipping...")
            continue
        dependency_path: str = NODE_MODULES_PATH + "/" + dependency_name
        try:
            print("Copying '" + dependency_path + "' to '" + OUTPUT_FOLDER_PATH + "node_modules/'")
            shutil.copytree(dependency_path, OUTPUT_FOLDER_PATH + "node_modules/" + dependency_name)
        except FileExistsError:
            print("Replacing existing dependency from '" + dependency_path + "' to '" + OUTPUT_FOLDER_PATH + "node_modules/'")
            shutil.rmtree(OUTPUT_FOLDER_PATH + "node_modules/" + dependency_name)
            time.sleep(0.1) # Add small delay
            shutil.copytree(dependency_path, OUTPUT_FOLDER_PATH + "node_modules/" + dependency_name)  
        
    
if __name__ == "__main__":
    main()

File: export.js
/*
Author: 
    aarontburn (https://github.com/aarontburn)
Title:
    Module Export Script
Purpose:
    Properly export developed module to use with parent application
        (https://github.com/aarontburn/modules)

Repository: 
    https://github.com/aarontburn/modules-export-script
    
Usage: 
    'npm run export', or 'node node_modules/modules-export-script/export.js'
    
Expected Result: 
    In the root directory, a directory 'output/' will be created containing required files for the module.
*/




const os = require('os');
const path = require("path");
const fs = require("fs");
const dialogNode = require('dialog-node')
const archiver = require('archiver')(os.platform() !== 'linux' ? 'zip' : 'tar');

// File name of the info file for the module.
const MODULE_INFO_FILE = "moduleinfo.json";

// The path of the root directory of this module.
const PWD = path.join(__dirname, "../", "../");

const FOLDER_NAME = (() => {
    const srcPath = path.join(PWD, "src");
    for (const file of fs.readdirSync(srcPath, { withFileTypes: true })) {
        if (file.isDirectory() && fs.readdirSync(path.join(file.path, file.name)).includes(MODULE_INFO_FILE)) {
            return file.name;
        }
    }
})();

if (FOLDER_NAME === undefined) {
    throw new Error(`Could not locate '${MODULE_INFO_FILE}'. Ensure your module folder contains it.`);
}

// The path of the output directory in the output folder
const _OUTPUT_FOLDER_PATH = PWD + "/output/" + FOLDER_NAME + "/";

// The path to the node_modules directory in the output folder.
const NODE_MODULES_PATH = PWD + "/node_modules";





let chosenFolder;

const getOutputFolder = () => chosenFolder === undefined ? _OUTPUT_FOLDER_PATH : chosenFolder;

async function main() {

    const outputPath = await getDirectory();
    console.log("Outputting module to: " + outputPath)
    if (outputPath !== undefined) {
        chosenFolder = outputPath;
    }

    fs.rmSync(getOutputFolder(), { recursive: true, force: true });
    modifyModuleInfoJSON();
    createDirectories();
    copyFiles();
    checkAndCopyDependencies();
    await toArchive();

    console.log("\n\tFINISHING BUNDLING MODULE");
}

async function getDirectory() {
    const promise = new Promise((resolve, _) => {
        dialogNode.fileselect('', 'Choose a folder to save your module in.', 0, (_, directory, __ ) => {
            if (directory === '') {
                resolve(undefined);
            }
            resolve(`${directory.trim()}/${FOLDER_NAME}/`);
        });
    })
    return promise;

} 

function modifyModuleInfoJSON() {
    const jsonPath = PWD + "/src/" + FOLDER_NAME + "/" + MODULE_INFO_FILE;
    const json = JSON.parse(fs.readFileSync(jsonPath));
    json["build_version"] += 1
    fs.writeFileSync(jsonPath, JSON.stringify(json));
}

function createDirectories() {
    function mkdir(directoryName) {
        fs.mkdirSync(directoryName, { recursive: true })
    }

    console.log("\n\tCREATING FOLDERS\n");
    mkdir(getOutputFolder());
    mkdir(getOutputFolder() + "module_builder");
    mkdir(getOutputFolder() + "node_modules");
}

function copyFiles() {
    console.log("\n\tCOPYING FILES\n");

    const dir = PWD + "/src/" + FOLDER_NAME + "/";
    for (const file of fs.readdirSync(dir, { withFileTypes: true })) {
        if (file.isDirectory() && file.name === "module_builder") {
            continue;
        }

        console.log(`Copying '${path.join(file.path, file.name)}' to output folder (${path.join(getOutputFolder(), file.name)})`);
        fs.cpSync(path.join(file.path, file.name), path.join(getOutputFolder(), file.name), { recursive: true });
    }
}



function checkAndCopyDependencies() {
    const json = JSON.parse(fs.readFileSync(PWD + "/package.json"));

    const dependencies = json["dependencies"];

    const dependencyNames = Object.keys(dependencies);
    if (dependencyNames.length > 1) {
        console.log("\n\tBUNDLING DEPENDENCIES\n");
    }


    const depSet = new Set()

    const nodeModules = fs.readdirSync(NODE_MODULES_PATH);

    for (const dependencyName of dependencyNames) {
        if (dependencyName === "modules-export-script") {
            continue
        }

        if (!nodeModules.includes(dependencyName)) {
            console.log(dependencyName + " was not found in 'node_modules'. Skipping...")
            continue;
        }

        depSet.add(dependencyName)
        checkDependencysDependencies(dependencyName, depSet)
    }

    console.log(depSet)


    depSet.forEach(depName => {
        const dependencyPath = path.join(NODE_MODULES_PATH, depName);
        console.log("Copying '" + dependencyPath + "' to '" + getOutputFolder() + "node_modules/'")
        fs.cpSync(dependencyPath, path.join(getOutputFolder(), "node_modules/" + depName), { recursive: true });
    })
}


function checkDependencysDependencies(depName, depSet) {
    const depDir = path.join(NODE_MODULES_PATH, depName);
    const depJson = path.join(depDir, "package.json");

    const json = JSON.parse(fs.readFileSync(depJson));
    const dependencies = json["dependencies"];

    if (dependencies !== undefined) {

        Object.keys(dependencies).forEach(name => {
            depSet.add(name);
            checkDependencysDependencies(name, depSet)
        })
    }
}

function toArchive() {
    const outputFolder = getOutputFolder();
    const stream = fs.createWriteStream(outputFolder.slice(0, -1) + '.zip');
    console.log("\n\tARCHIVING FOLDER")
    console.log(`From ${outputFolder} to ${outputFolder.slice(0, -1)}.zip`);
    return new Promise((resolve, reject) => {
        archiver
          .directory(outputFolder, false)
          .on('error', err => reject(err))
          .pipe(stream)
        ;
    
        stream.on('close', () => resolve());
        archiver.finalize();
      });


}





main();